### PR TITLE
Release 3.0.0

### PR DESCRIPTION
**Pull request by Mo-Lala84**
### JIRA link (if applicable) ###
RHF
SHC
Accessibility

### Change description ###
This pr contains the changes to support 22.1(rhf, show hide condition and accessibility fixes)

An aggregation of a number of other PRs, each of which has been approved.

**Does this PR introduce a breaking change?** (check one with "x")

```
[x] Yes, which is why it's going up to version 4.0.0
[ ] No
```

### CODE REVIEW

**Review by udaydenduluri33 — DISMISSED** [review]
LGTMm

File: src/shared/directives/conditional-show/conditional-show-form.directive.ts [hunk 1]
```diff
@@ -95,16 +88,11 @@ export class ConditionalShowFormDirective implements OnInit, AfterViewInit, OnDe
           cf.hidden = false;
         }
         if (condResult === cf.hidden) {
-          if (cf.hidden) {
-            this.greyBarService.addToggledToShow(cf.id);
-          } else {
-            this.greyBarService.removeToggledToShow(cf.id)
-          }
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/conditional-show-form.directive.ts:102]: hope there is reason to remove this logic

File: src/shared/directives/conditional-show/conditional-show-form.directive.ts [hunk 2]
```diff
@@ -121,18 +109,18 @@ export class ConditionalShowFormDirective implements OnInit, AfterViewInit, OnDe
...
-  private handleFormControl = (c: FormControl) => {
+  private handleFormControl = (c: FormControl): void => {
     this.evaluateControl(c);
   }
 
-  private handleFormArray = (c: FormArray, caseField: CaseField) => {
-    this.evaluateControl(c);
-    c.controls.forEach((formControl, ix) => {
-      this.fieldsUtils.controlIterator(formControl, this.handleFormArray, this.handleFormGroup, this.handleFormControl)
+  private handleFormArray = (a: FormArray): void => {
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/conditional-show-form.directive.ts:116]: I'm not a fan to use single letter as obj reference, but if we amend it we could rename it to something meaningful, formArray?

File: src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts [hunk 3]
```diff
@@ -263,11 +376,19 @@ export class CaseEditSubmitComponent implements OnInit, OnDestroy {
...
-  getCaseId(): String {
+  private getCaseFields(): CaseField[] {
+    if (this.caseEdit.caseDetails) {
+      return FieldsUtils.getCaseFields(this.caseEdit.caseDetails);
+    }
+
+    return this.eventTrigger.case_fields;
+  }
+
+  public getCaseId(): String {
```
Comment (rayliangatos) [inline src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts:387]: should this be string but not the String wrapper?

File: src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts [hunk 4]
```diff
@@ -263,11 +376,19 @@ export class CaseEditSubmitComponent implements OnInit, OnDestroy {
...
     return (this.caseEdit.caseDetails ? this.caseEdit.caseDetails.case_id : '');
   }
 
-  getCancelText(): String {
+  public getCancelText(): String {
```
Comment (rayliangatos) [inline src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts:391]: should this be string but not the String wrapper?

File: src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts [hunk 5]
```diff
@@ -119,11 +130,114 @@ export class CaseEditSubmitComponent implements OnInit, OnDestroy {
...
+          // handle both null and undefined values
+          case 'Complex':
+            if (caseField.retain_hidden_value && caseField.value != null) {
+              // Call this function recursively to replace the Complex field's sub-fields as necessary, passing the
+              // CaseField itself (the sub-fields do not contain any values, so these need to be obtained from the
+              // parent)
+              const resultantObject = this.replaceHiddenFormValuesWithOriginalCaseData(
+                formGroup.controls[key] as FormGroup, caseField.field_type.complex_fields, caseField);
+              // Update rawFormValueData for this field
+              rawFormValueData[key] = resultantObject;
```
Comment (rayliangatos) [inline src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts:196]:
```suggestion
              rawFormValueData[key] = this.replaceHiddenFormValuesWithOriginalCaseData(
                formGroup.controls[key] as FormGroup, caseField.field_type.complex_fields, caseField);
```

**Comment by rayliangatos** [comment]
> SonarCloud Quality Gate failed.
> 
> [Bug](https://sonarcloud.io/project/issues?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=VULNERABILITY)
> [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=SECURITY_HOTSPOT) [E](https://sonarcloud.io/project/security_hotspots?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=SECURITY_HOTSPOT) [12 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=SECURITY_HOTSPOT)
> [Code Smell](https://sonarcloud.io/project/issues?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=CODE_SMELL) [92 Code Smells](https://sonarcloud.io/project/issues?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&resolved=false&types=CODE_SMELL)
> 
> [No Coverage information](https://sonarcloud.io/component_measures?id=hmcts_ccd-case-ui-toolkit&pullRequest=818) No Coverage information
> [](https://sonarcloud.io/component_measures?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&metric=new_duplicated_lines_density&view=list) [77.0% Duplication](https://sonarcloud.io/component_measures?id=hmcts_ccd-case-ui-toolkit&pullRequest=818&metric=new_duplicated_lines_density&view=list)

Though many of the code smell is from the spec.ts, it worth to check the codesmell of these files:
src/.../case-editor/case-edit-submit/case-edit-submit.component.ts 
src/shared/services/fields/fields.purger.ts
src/shared/services/form/form-value.service.ts

File: src/case-ui-toolkit.module.ts [hunk 6]
```diff
@@ -78,7 +77,6 @@ import { CaseHistoryModule } from './shared/components/case-history';
       CaseViewComponent,
       CaseViewerComponent,
       CallbackErrorsComponent,
-      PaletteModule,
```
[the remark this answers is not in the document]
Comment (Daniel-Lam) [reply src/case-ui-toolkit.module.ts:81]: @rayliangatos - `PaletteModule` was removed from `exports` by @phillip-whitaker in commit 9310a61 to `master`.

File: src/shared/directives/conditional-show/conditional-show-form.directive.ts (hunk 1 above)
Comment (Daniel-Lam) [reply src/shared/directives/conditional-show/conditional-show-form.directive.ts:102]: @rayliangatos - this logic was deleted by @paul-graham in commit ebea39b in PR #764

Comment (paul-graham) [reply src/shared/directives/conditional-show/conditional-show-form.directive.ts:102]: @rayliangatos, @Daniel-Lam:

Yes, the `GreyBarService` is no longer used for displaying the grey bars - an entirely different approach is in place. This was done as part of https://tools.hmcts.net/jira/browse/EUI-3267 and some tidying up happened subsequently in https://tools.hmcts.net/jira/browse/EUI-3854.

File: src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts [hunk 7]
```diff
@@ -83,24 +84,34 @@ export class CaseEditSubmitComponent implements OnInit, OnDestroy {
...
+  public submit(): void {
     this.isSubmitting = true;
-    let caseEventData: CaseEventData = this.formValueService.sanitise(this.editForm.value) as CaseEventData;
+    const caseEventData: CaseEventData = {
+      data: this.replaceEmptyComplexFieldValues(
+        this.formValueService.sanitise(
+          this.replaceHiddenFormValuesWithOriginalCaseData(
+            this.editForm.get('data') as FormGroup, this.eventTrigger.case_fields))),
+      event: this.editForm.value.event
+    } as CaseEventData
```
[the remark this answers is not in the document]
Comment (Daniel-Lam) [reply src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts:102]: Added missing `;`

File: src/shared/components/case-editor/case-edit-page/case-edit-page.html [hunk 8]
```diff
@@ -64,7 +79,7 @@ <h3 class="heading-h3 error-summary-heading" id="edit-case-event_error-summary-h
       <button class="button button-secondary" type="button" [disabled]="!(hasPreviousPage$ | async)" (click)="toPreviousPage()">
         Previous
       </button>
-      <button class="button" type="submit" [disabled]="currentPageIsNotValid() || submitting()">{{triggerText}}</button>
+      <button class="button" type="submit" [disabled]="submitting()">{{triggerText}}</button>
```
[the remark this answers is not in the document]
Comment (Daniel-Lam) [reply src/shared/components/case-editor/case-edit-page/case-edit-page.html:82]: This change was not introduced as part of the "retain hidden fields" work.

File: src/shared/components/case-viewer/case-viewer.component.html [hunk 9]
```diff
@@ -52,7 +52,7 @@ <h2 class="heading-h2 error-summary-heading" id="edit-case-event_error-summary-h
             <ng-template matTabContent>
               <table [class]="tab.id">
                 <tbody>
-                <ng-container *ngFor="let field of tab | ccdTabFields | ccdReadFieldsFilter:false :undefined :true">
+                <ng-container *ngFor="let field of tab | ccdTabFields | ccdReadFieldsFilter:false :undefined :true :formGroup.controls['data']">
```
[the remark this answers is not in the document]
Comment (Daniel-Lam) [reply src/shared/components/case-viewer/case-viewer.component.html:55]: This change was not introduced as part of the "retain hidden fields" work.

File: src/shared/components/palette/collection/write-collection-field.html [hunk 10]
```diff
@@ -2,13 +2,13 @@
 
   <div class="panel collection-indicator">
 
-    <h2 class="heading-h2">
+    <h2 *ngIf="caseField.hint_text || formArray.errors" class="heading-h2 error-spacing">
```
[the remark this answers is not in the document]
Comment (Daniel-Lam) [reply src/shared/components/palette/collection/write-collection-field.html:5]: This change was not introduced as part of the "retain hidden fields" work.

File: src/shared/components/palette/phone-uk/write-phone-uk-field.html [hunk 11]
```diff
@@ -1,10 +1,11 @@
...
+<div class="form-group" [ngClass]="{'form-group-error': !phoneUkControl.valid && phoneUkControl.dirty}">
 
   <label [for]="id()">
     <span class="form-label" *ngIf="caseField.label">{{caseField | ccdFieldLabel}}</span>
-    <span class="form-hint" *ngIf="caseField.hint_text">{{caseField.hint_text}}</span>
-    <span class="error-message" *ngIf="phoneUkControl.errors && phoneUkControl.touched">{{phoneUkControl.errors | ccdFirstError}}</span>
   </label>
-  <input class="form-control bottom-30" [id]="id()" type="tel" [formControl]="phoneUkControl">
+  <span class="form-hint" *ngIf="caseField.hint_text">{{caseField.hint_text}}</span>
+  <span class="error-message" *ngIf="phoneUkControl.errors && phoneUkControl.dirty">{{phoneUkControl.errors | ccdFirstError:caseField.label}}</span>
```
[the remark this answers is not in the document]
Comment (Daniel-Lam) [reply src/shared/components/palette/phone-uk/write-phone-uk-field.html:7]: This change was not introduced as part of the "retain hidden fields" work.

File: src/shared/components/palette/text-area/write-text-area-field.html [hunk 12]
```diff
@@ -1,14 +1,12 @@
...
-<div class="form-group" [ngClass]="{'form-group-error': !textareaControl.valid && textareaControl.touched}">
+<div class="form-group" [ngClass]="{'form-group-error': !textareaControl.valid && textareaControl.dirty}">
 
   <label [for]="id()">
     <span class="form-label" *ngIf="caseField.label">{{caseField | ccdFieldLabel}}</span>
-    <span class="form-hint" *ngIf="caseField.hint_text">{{caseField.hint_text}}</span>
-    <span class="error-message" *ngIf="textareaControl.errors && textareaControl.touched">{{textareaControl.errors | ccdFirstError}}</span>
   </label>
+  <span class="form-hint" *ngIf="caseField.hint_text">{{caseField.hint_text}}</span>
+  <span class="error-message" *ngIf="textareaControl.errors && textareaControl.dirty">{{textareaControl.errors | ccdFirstError:caseField.label}}</span>
```
[the remark this answers is not in the document]
Comment (Daniel-Lam) [reply src/shared/components/palette/text-area/write-text-area-field.html:7]: This change was not introduced as part of the "retain hidden fields" work.

File: src/shared/components/palette/datetime-picker/datetime-picker.component.ts [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+      if (this.formatTranslationService.hasMinutes(dateTimePickerFormat)) {
+        this.hideMinutes = false;
+        this.disableMinute = false;
+        this.hideTime = false;
+        if (!this.formatTranslationService.is24Hour(dateTimePickerFormat)) {
+          this.enableMeridian = true;
+        }
+        return;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (Daniel-Lam) [reply src/shared/components/palette/datetime-picker/datetime-picker.component.ts:147]: This change was not introduced as part of the "retain hidden fields" work.

File: src/shared/components/palette/text/write-text-field.html [hunk 14]
```diff
@@ -1,10 +1,12 @@
...
+<div class="form-group" [ngClass]="{'form-group-error': !textControl.valid && textControl.dirty}">
 
   <label [for]="id()">
     <span class="form-label" *ngIf="caseField.label">{{caseField | ccdFieldLabel}}</span>
-    <span class="form-hint" *ngIf="caseField.hint_text">{{caseField.hint_text}}</span>
-    <span class="error-message" *ngIf="textControl.errors && textControl.touched">{{textControl.errors | ccdFirstError}}</span>
   </label>
-  <input class="form-control bottom-30" [id]="id()" type="text" [formControl]="textControl">
+  <span class="form-hint" *ngIf="caseField.hint_text">{{caseField.hint_text}}</span>
+  <span class="error-message" *ngIf="textControl.errors && textControl.dirty">{{textControl.errors | ccdFirstError:caseField.label}}</span>
```
[the remark this answers is not in the document]
Comment (Daniel-Lam) [reply src/shared/components/palette/text/write-text-field.html:7]: This change was not introduced as part of the "retain hidden fields" work.

File: src/shared/directives/conditional-show/conditional-show-form.directive.ts (hunk 2 above)
Comment (Daniel-Lam) [reply src/shared/directives/conditional-show/conditional-show-form.directive.ts:116]: This change was not introduced as part of the "retain hidden fields" work, but I can rename it to `formArray`.

Comment (Daniel-Lam) [reply src/shared/directives/conditional-show/conditional-show-form.directive.ts:116]: I can do the same for `handleFormControl` and `handleFormGroup` functions.

File: src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts (hunk 3 above)
Comment (Daniel-Lam) [reply src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts:387]: Changed return type to `string`.

File: src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts (hunk 4 above)
Comment (Daniel-Lam) [reply src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts:391]: Changed return type to `string`.

File: src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts (hunk 5 above)
Comment (Daniel-Lam) [reply src/shared/components/case-editor/case-edit-submit/case-edit-submit.component.ts:196]: Refactored as per suggestion.

**Review by paul-graham — APPROVED** [review]
Pretty much the whole team have been involved in this one. My part is fairly small and I've approved many of the other PRs that have ended up in here and it's all been regression tested... so I'm just going to approve this.

**Review by udaydenduluri33 — APPROVED** [review]
LGTMm
These changes are done as a part of next major release. I understand each change is a separate PR thats merged into release branch.

**Pull request merged**